[WIP] Move NETFramework ns2.0 support into SDK

ericstj opened the pull request:
/cc @dsplaisted 

This is a work in progress.

ericstj committed:
Add task for determining if NETStandard is referenced

ericstj committed:
Add a separate build of tasks for NETFramework

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.net46.cs:
```diff
@@ -0,0 +1,163 @@
+        internal static bool GetFileDependsOnNETStandard(string filePath)
+        {
+            // Ported from Microsoft.Build.Tasks.AssemblyInformation
+            if (Environment.OSVersion.Platform == PlatformID.Unix || Environment.OSVersion.Platform == PlatformID.MacOSX)
```
Is there going to be any way for us to test this Mono logic?

ericstj replied:
I tested it manually in debugger.  I could provide a hook for testing it on Windows.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.net46.cs:
```diff
@@ -0,0 +1,163 @@
+            {
+                // on Windows use CLR's unmanaged metadata API.
+                // Create the metadata dispenser and open scope on the source file.
+                IMetaDataDispenser metadataDispenser = (IMetaDataDispenser)new CorMetaDataDispenser();
```
We have a lot of our full framework coverage disabled due to #1041 (Jenkins needs to be updated).  We should either fix that in order to get decent coverage of this or at least make sure the tests pass locally.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.NETFramework/Microsoft.NET.Build.Tasks.NETFramework.csproj:
```diff
@@ -0,0 +1,111 @@
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <PackageId>Microsoft.NET.Sdk</PackageId>
+    <Description>The MSBuild targets and properties for building .NET Core projects.</Description>
```
This Description isn't accurate.  A lot of the metadata can probably be deleted, and I don't think we need to multi-target this project.

ericstj replied:
I'll remove some of the goo from this project.  Do you think it even needs to have a package, or could we just put this stuff in some single package.

dsplaisted replied:
I think it can all go into the same package.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.NETFramework/Microsoft.NET.Build.Tasks.NETFramework.csproj:
```diff
@@ -0,0 +1,111 @@
+
+  <ItemGroup>
+    <Compile Include="..\Microsoft.NET.Build.Tasks\BuildErrorException.cs">
+      <Link>BuildErrorException.cs</Link>
```
Setting the `Link` metadata explicitly isn't required as of #1246.  And for the ConflictResolution files, you can just set `LinkBase="ConflictResolution"`.  I'm not sure whether that change has been integrated into the version of the CLI that the SDK repo uses to build, though.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.NETFramework/Microsoft.NET.Build.Tasks.NETFramework.csproj:
```diff
@@ -0,0 +1,111 @@
+      <Link>ItemUtilities.cs</Link>
+    </Compile>
+    <Compile Include="..\Microsoft.NET.Build.Tasks\GetDependsOnNETStandard*.cs">
+      <Link>%(FileName)%(Exetnsion)</Link>
```
Typo: `Exetnsion` (but as mentioned previously, this shouldn't be needed at all, at least once the update flows to the right place)

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.NETFramework/Microsoft.NET.Build.Tasks.NETFramework.csproj:
```diff
@@ -0,0 +1,111 @@
+
+  </ItemGroup>
+
+  <Target Name="PrepareAdditionalFilesToLayout" BeforeTargets="AssignTargetPaths">
```
Instead of having any of the package layout stuff in this project, we probably want building the other project to include the layout from this one.

ericstj committed:
Add NETStandard.Library.NETFramework content

ericstj committed:
Condition inclusion of NS2 netfx facades.

ericstj commented:
/cc @livarcocc since @dsplaisted is out next week.  We'll probably need someone else to review this.

Here's the work I have  left to do:
1.	Cross-compile the task dll, only use that task DLL and don't try to reuse SDK DLL.
2.	Use version comparisons eg: '$(_TargetFrameworkVersionWithoutV)' >= '4.7'
3.	Add a conflict resolution target after MS.Common.targets
a.	Needs to use a different name: Microsoft.NET.Build.Extensions.
b.	Enable conflict resolution by default for NETFx, opt-in otherwise.
4.	Delete conflict resolution from standard repo.

ericstj committed:
Support override of targets file location

ericstj committed:
Use TargetFrameworkVersion comparison, check for inbox

Future-proof this target by turning off the functionality once netstandard.dll is inbox.

Improve conditions by doing version comparison instead of string equality checks.
